harness/determined/searcher/search_runner.py

```diff
@@ -0,0 +1,243 @@
+                            )
+                        request_id = uuid.UUID(event.trialExitedEarly.requestId)
+                        if event.trialExitedEarly.exitedReason in (
+                            v1TrialExitedEarlyExitedReason.EXITED_REASON_INVALID_HP,
+                            v1TrialExitedEarlyExitedReason.EXITED_REASON_INIT_INVALID_HP,
+                        ):
+                            self.search_method.searcher_state.trial_progress.pop(request_id, None)
+                        elif (
+                            event.trialExitedEarly.exitedReason
+                            == v1TrialExitedEarlyExitedReason.EXITED_REASON_UNSPECIFIED
+                        ):
+                            self.search_method.searcher_state.failures.add(request_id)
+                        operations = self.search_method.on_trial_exited_early(
+                            request_id,
+                            exited_reason=ExitedReason._from_bindings(
+                                event.trialExitedEarly.exitedReason
+                            ),
+                        )
+                    elif event.validationCompleted:
+                        # duplicate completion accounting already performed by master
+                        logging.info(
+                            f"validationCompleted({event.validationCompleted.requestId},"
+                            f" {event.validationCompleted.metric})"
+                        )
+                        request_id = uuid.UUID(event.validationCompleted.requestId)
+                        if event.validationCompleted.metric is None:
+                            raise RuntimeError(
+                                "validationCompleted event is invalid without a metric"
+                            )
+                        operations = self.search_method.on_validation_completed(
+                            request_id,
+                            event.validationCompleted.metric,
+                        )
+                    elif event.experimentInactive:
+                        logging.info(
+                            f"experiment {self.search_method.searcher_state.experiment_id} is "
+                            f"inactive; state={event.experimentInactive.experimentState}"
+                        )
+                        experiment_is_active = False
+                        break
+                    elif event.trialProgress:
+                        logging.debug(
+                            f"trialProgress({event.trialProgress.requestId}, "
+                            f"{event.trialProgress.partialUnits})"
+                        )
+                        request_id = uuid.UUID(event.trialProgress.requestId)
+                        self.search_method.searcher_state.trial_progress[request_id] = float(
+                            event.trialProgress.partialUnits
+                        )
```

Comment:
Who owns the SearcherState?

It seems to have some mixed ownership, which is usually a messy experience.  There are some things that the user needs to be worried about some things that are populated automatically (like trial progress) which the user should consider read-only.

Is there a way to draw a clearer line between what is owned by the system vs what is owned by the user code?

Reply:
Searcher owns it but we want to make it available to the search method. As we discussed offline, we could emphasize that the user should not modify it by passing it to event handlers as a parameter.